src/main.cpp

```diff
@@ -3738,6 +3759,108 @@ bool CVerifyDB::VerifyDB(CCoinsView *coinsview, int nCheckLevel, int nCheckDepth
     return true;
 }
 
+bool RewindBlockIndex(const CChainParams& params)
+{
+    LOCK(cs_main);
+
+    // RewindBlockIndex is called after LoadBlockIndex, so at this point every block
+    // index will have nBranchId set either to the persisted value, or the value of
+    // its parent block. Thus we can reason here that for valid chains:
+    //
+    // - BLOCK_OPT_ACTIVATES_NU must be set only on blocks that activate upgrades.
+    // - nBranchId for each block must match what we expect.
+    auto sufficientlyValidated = [&params](const CBlockIndex* pindex) {
+        auto consensus = params.GetConsensus();
+        bool fFlagSet = pindex->nStatus & BLOCK_OPT_ACTIVATES_NU;
+        bool fFlagExpected = IsActivationHeight(pindex->nHeight, consensus);
+        return fFlagSet == fFlagExpected && pindex->nBranchId == CurrentEpochBranchId(pindex->nHeight, consensus);
+    };
+
+    int nHeight = 1;
+    while (nHeight <= chainActive.Height()) {
+        if (!sufficientlyValidated(chainActive[nHeight])) {
+            break;
+        }
+        nHeight++;
+    }
+
+    // nHeight is now the height of the first insufficiently-validated block, or tipheight + 1
```

Comment:
I will rebase #2463 on top of this PR, and add a restriction here to prevent rollbacks past the limit.